### DISCUSSION TITLE
Folder already exists

### DISCUSSION THREAD

**Discussion by vikashg** [Project-MONAI/monai-deploy-app-sdk#427] [General]
Hi, 
I am getting this error 
![image](https://github.com/Project-MONAI/monai-deploy-app-sdk/assets/3863212/6504b88c-ffb0-4ee4-875a-1ad6d4c54f9d)
It is originating from my segmentation operator which I wrote following livertumor_seg_operator
![image](https://github.com/Project-MONAI/monai-deploy-app-sdk/assets/3863212/5896fdba-76c8-44a5-9c50-844f01b6c5ba)

I am not able to figure why ?
@MMelQin Thoughts. Let me know if you need more code to answer this.

Thanks 
Vikash

**Comment by MMelQin** [comment]
> Hi, I am getting this error ![image](https://user-images.githubusercontent.com/3863212/239074441-6504b88c-ffb0-4ee4-875a-1ad6d4c54f9d.png) It is originating from my segmentation operator which I wrote following livertumor_seg_operator ![image](https://user-images.githubusercontent.com/3863212/239074659-5896fdba-76c8-44a5-9c50-844f01b6c5ba.png)
> 
> I am not able to figure why ? @MMelQin Thoughts. Let me know if you need more code to answer this.
> 
> Thanks Vikash

Hi @vikashg, the code shows that it is using the v0.5 of App SDK. The specific section of the code is similar to [the one](https://github.com/Project-MONAI/monai-deploy-app-sdk/blob/f62199b44419bf3e86a132bf92f2f09db3940034/examples/apps/ai_livertumor_seg_app/livertumor_seg_operator.py#LL35C13-L35C32) in the Liver Tumor app, The exception says the output folder, `save_images_folder`, already exists, but from the code snippet, I cannot tell how that could be the case. An operator instance is identified by a GUID, and this GUID is used as the folder name for the on disk storage of the operator instance.

I will try to reproduce the issue with the liver example, from the code in the main branch, or from the [Docker image](ghcr.io/mmelqin/monai_ai_livertumor_seg_app:1.0)

**Reply by vikashg** [reply]
Thanks Ming.
I just cloned the repo and switched to your branch mqin/sdk_migration. Here is the gist 
https://gist.github.com/vikashg/53733780de8d2a7afa90c5aadb506792
 I will look at it again today. See what's going on

**Reply by MMelQin** [reply]
If using the mqin/sdk_migration branch, then the operator and app will need to be using the WIP new App SDK.

In any case, a bit about this specific output, `@md.output("saved_images_folder", DataPath, IOType.DISK)`, and see if it is needed in the app you are building
- With the ClaraViz integration operator directly consuming the in_memory image objects from the inference operator etc, there is no need for this particular output. Before that, I had to use directly user ClaraViz's API, which expects both the volume image files of the density and segmentation, along with a metadata. So, the `save_image_folder` was needed to save the image files within the inference operator, and this data path is then passed to the `Publishing` operator to package them into the data input for ClaraViz.
- In the new WIP App SDK, using DISK as an operator's named output is NOT supported at all (dur to Holoscan yet to support this use case). So, the output folder has to be set in an operator's constructor, or the path values can be passed to an operator in its named input as str type. Similar, the convey the info to the downstream operators, the path value can be emitted in a named output as str type. There are examples in the branch for SDK_migration.

Thanks.

**Comment by mpsampat** [comment]
@vikashg , @MMelQin we had this same error with 0.5.1 about ~2 months ago. 
I will try and reproduce it and post back here.

**Reply by MMelQin** [reply]
The error was actually thrown by the IO context itself, not from the OS/FS. It is limitation, as designed, that during the operator output setup, the IO context will disallow the output be DISK type if the operator being created is an interior operator; only leaf operator in a DAG can have DISK output.

When files/folders need to be created by an interior operator, for saving intermediate results to disk, AND need to pass the path(s) to the downstream operator, the folder/file paths need to be created under the Application output folder, so that all operators can locate the parent folder via the app context. The paths then need to be wrapped in an object, which is then used to set the operator output of the IM_MEMORY type. The downstream operator/consumer then needs to the corresponding input of the in_memory object type, as well as the logic to parse it.

In MD App SDK, there is completely no support for DISK in input, output, or app execution context. I had to come up with the recipe/design to pass the folder on the initializers and still need to resort to the in memory object in the operator output to pass the paths to consuming operators.
